File: src/cee/app/local_llm/dream_consolidator.py
"""
认知涌现引擎 — 梦境巩固器
=========================
模拟人脑睡眠期的知识巩固与创意重组

四个梦境模式:
1. 记忆回放(replay): 按时间顺序回放最近的对话片段
   → 高频调用的事实增加置信度；被遗忘的(长期未召回)轻微衰减

2. 随机融合(shuffle): 随机配对两个不同领域的事实
   → 通过ConceptualBlending产生惊喜连接，存入"潜洞见"队列
   → 潜洞见在后续对话中若被验证(用户like)则升级为正式知识

3. 抽象爬梯(ladder): 对同一话题的事实向上抽取出共性规律
   → 例如: "神经网络用梯度下降"+"决策树用分裂增益" → "机器学习算法都依赖优化准则"
   → 生成更高层级的抽象知识

4. 异常侦测(anomaly): 检测知识库中的矛盾事实
   → "Python是编译型语言"(错) vs "Python是解释型语言"(对)
   → 标记低置信度矛盾，触发人类确认

触发条件:
- consolidate()被调用时执行一次完整的睡眠周期
- 建议在闲置期(如无用户请求>30秒)或每50轮对话触发

用法:
    dc = DreamConsolidator(knowledge_store, auto_learner, kgraph)
    results = dc.dream_cycle()
    # -> {"consolidated": 3, "blended": 1, "abstracted": 2, "anomalies": 1}
"""

from __future__ import annotations

import json
import random
import re
from collections import Counter
from dataclasses import dataclass, field
from datetime import datetime, timezone, timedelta
from pathlib import Path
from typing import TYPE_CHECKING

from .knowledge_store import STORAGE_DIR

if TYPE_CHECKING:
    from .knowledge_store import SelfLearningKnowledgeStore
    from .auto_learner import AutoLearner
    from .knowledge_graph import KnowledgeGraph
    from .conceptual_blending import ConceptualBlending

DREAM_FILE = STORAGE_DIR / "dream_consolidator.json"


@dataclass
class DreamInsight:
    id: str
    dream_type: str
    content: str
    source_facts: list[str]
    confidence: float
    verified: bool = False
    created_at: str = field(default_factory=lambda: datetime.now(timezone.utc).isoformat())


class DreamConsolidator:
    def __init__(self, knowledge_store: SelfLearningKnowledgeStore | None = None,
                 auto_learner: AutoLearner | None = None,
                 kgraph: KnowledgeGraph | None = None,
                 blender: ConceptualBlending | None = None):
        self._store = knowledge_store
        self._learner = auto_learner
        self._kgraph = kgraph
        self._blender = blender

        self._insights: list[DreamInsight] = []
        self._dream_count: int = 0
        self._last_dream: str = datetime.now(timezone.utc).isoformat()
        self._cycles: list[dict] = []
        self._load()

    def dream_cycle(self) -> dict:
        self._dream_count += 1
        results = {"consolidated": 0, "blended": 0, "abstracted": 0,
                   "anomalies": 0, "emergent_insights": []}

        self._memory_replay()
        results["consolidated"] = self._consolidate_facts()

        if self._learner:
            facts = self._learner._facts
            if len(facts) >= 4:
                blend_count = self._shuffle_blend(facts)
                results["blended"] = blend_count

            if len(facts) >= 6:
                abstracted = self._ladder_abstraction(facts)
                results["abstracted"] = abstracted

            if len(facts) >= 5:
                anomalies = self._detect_anomalies(facts)
                results["anomalies"] = anomalies

        if self._kgraph and self._dream_count % 5 == 0:
            removed = self._kgraph.decay_edges(decay_factor=0.97, min_weight=0.15)
            results["kg_edge_decay"] = removed

        self._cycles.append(results)
        if len(self._cycles) > 50:
            self._cycles = self._cycles[-30:]

        self._last_dream = datetime.now(timezone.utc).isoformat()
        self._save()
        return results

    def verify_insight(self, insight_id: str, verified: bool = True):
        for ins in self._insights:
            if ins.id == insight_id:
                ins.verified = verified
                ins.confidence = min(1.0, ins.confidence + 0.3) if verified else max(0.2, ins.confidence - 0.2)
                if verified and self._learner:
                    self._learner._stats["learned"] += 1
                self._save()
                return

    def get_pending_insights(self, top_k: int = 3) -> list[DreamInsight]:
        unverified = [i for i in self._insights if not i.verified]
        unverified.sort(key=lambda x: x.confidence, reverse=True)
        return unverified[:top_k]

    def _memory_replay(self):
        pass

    def _consolidate_facts(self) -> int:
        if self._learner:
            return self._learner.consolidate()
        return 0

    def _shuffle_blend(self, facts: list) -> int:
        if not self._blender or len(facts) < 4:
            return 0

        blended = 0
        by_topic: dict[str, list] = {}
        for f in facts:
            by_topic.setdefault(f.topic, []).append(f)

        topics = list(by_topic.keys())
        if len(topics) < 2:
            return 0

        for _ in range(min(3, len(topics) - 1)):
            t1, t2 = random.sample(topics, 2)
            f1 = random.choice(by_topic[t1])
            f2 = random.choice(by_topic[t2])

            a_name = self._extract_core_concept(f1.fact)
            b_name = self._extract_core_concept(f2.fact)

            if a_name and b_name and a_name != b_name:
                result = self._blender.blend(a_name, b_name, blend_type="auto")
                if result and result.score > 0.55:
                    insight = DreamInsight(
                        id=f"dream_{self._dream_count}_{blended + 1}",
                        dream_type="shuffle",
                        content=result.insight,
                        source_facts=[f1.fact, f2.fact],
                        confidence=result.score,
                    )
                    self._insights.append(insight)
                    blended += 1

        if len(self._insights) > 200:
            self._insights = self._insights[-100:]

        return blended

    def _ladder_abstraction(self, facts: list) -> int:
        by_topic: dict[str, list] = {}
        for f in facts:
            by_topic.setdefault(f.topic, []).append(f)

        abstracted = 0
        for topic, fact_list in by_topic.items():
            if len(fact_list) < 3:
                continue
            text = " ".join(f.fact for f in fact_list[-6:])

            keywords = re.findall(r'[\u4e00-\u9fa5a-zA-Z]{2,12}', text)
            freq = Counter(keywords)
            common = [kw for kw, cnt in freq.most_common(5) if cnt >= 2 and len(kw) > 1]

            if len(common) >= 2:
                core = common[0]
                pattern = common[1]
                abstraction = f"在{topic}领域，{core}与{pattern}构成核心机制，"
                modality = self._infer_modality(abstraction, fact_list)
                if modality:
                    abstraction += f"共有的底层模式是{modality}"

                insight = DreamInsight(
                    id=f"ladder_{self._dream_count}_{abstracted + 1}",
                    dream_type="ladder",
                    content=abstraction,
                    source_facts=[f.fact for f in fact_list[:4]],
                    confidence=0.55,
                )
                self._insights.append(insight)
                abstracted += 1

        return abstracted

    def _detect_anomalies(self, facts: list) -> int:
        anomalies = 0
        for i, f1 in enumerate(facts):
            for f2 in facts[i + 1:]:
                if f1.topic == f2.topic and self._are_contradictory(f1.fact, f2.fact):
                    insight = DreamInsight(
                        id=f"anomaly_{self._dream_count}_{anomalies + 1}",
                        dream_type="anomaly",
                        content=f"潜在矛盾: [{f1.fact}] vs [{f2.fact}] — 需人工确认",
                        source_facts=[f1.fact, f2.fact],
                        confidence=0.35,
                    )
                    self._insights.append(insight)
                    anomalies += 1
                if anomalies >= 5:
                    break
            if anomalies >= 5:
                break
        return anomalies

    def _extract_core_concept(self, fact: str) -> str:
        match = re.match(r'([\u4e00-\u9fa5a-zA-Z]{2,16})', fact)
        if match:
            return match.group(1).strip()

        known = [
            "认知涌现", "神经网络", "深度学习", "软件架构", "人脑",
            "生物进化", "语言", "数学", "代码编译器", "免疫系统",
            "金融市场", "混沌边缘", "Python", "认知几何",
        ]
        for k in known:
            if k in fact:
                return k
        return ""

    def _infer_modality(self, text: str, facts: list) -> str:
        patterns = {
            "优化": "基于目标函数的最小化原则",
            "学习": "从经验中归纳出可泛化的规则",
            "结构": "层级组织的模块化设计",
            "演化": "变异+选择的迭代过程",
            "计算": "符号转换与状态迁移",
            "信息": "不确定性的消除与传递",
            "网络": "节点间的连接与信号传播",
        }
        for kw, mod in patterns.items():
            if kw in text:
                return mod
        return ""

    def _are_contradictory(self, f1: str, f2: str) -> bool:
        negation_pairs = [
            ("是", "不是"), ("可以", "不可以"), ("能", "不能"),
            ("属于", "不属于"), ("支持", "不支持"),
        ]
        for pos, neg in negation_pairs:
            if pos in f1 and neg in f2:
                return True
            if neg in f1 and pos in f2:
                return True
        return False

    def stats(self) -> dict:
        return {
            "dream_count": self._dream_count,
            "insights_total": len(self._insights),
            "insights_unverified": sum(1 for i in self._insights if not i.verified),
            "last_dream": self._last_dream,
            "recent_cycles": self._cycles[-5:] if self._cycles else [],
        }

    def _load(self):
        if DREAM_FILE.exists():
            try:
                with open(DREAM_FILE, "r", encoding="utf-8") as f:
                    data = json.load(f)
                self._insights = [DreamInsight(**i) for i in data.get("insights", [])[-100:]]
                self._dream_count = data.get("dream_count", 0)
                self._last_dream = data.get("last_dream", datetime.now(timezone.utc).isoformat())
                self._cycles = data.get("cycles", [])[-30:]
            except Exception:
                pass

    def _save(self):
        try:
            data = {
                "insights": [{"id": i.id, "dream_type": i.dream_type,
                              "content": i.content, "source_facts": i.source_facts,
                              "confidence": i.confidence, "verified": i.verified,
                              "created_at": i.created_at}
                             for i in self._insights[-100:]],
                "dream_count": self._dream_count,
                "last_dream": self._last_dream,
                "cycles": self._cycles[-30:],
                "updated_at": datetime.now(timezone.utc).isoformat(),
            }
            with open(DREAM_FILE, "w", encoding="utf-8") as f:
                json.dump(data, f, ensure_ascii=False, indent=2)
        except Exception:
            pass

File: middle-end/src/cee/app/local_llm/dream_consolidator.ts
/**
 * 梦境巩固器 — TypeScript 端
 *
 * 模拟睡眠期知识巩固：记忆回放、随机融合、抽象爬梯、异常侦测。
 * 从评估结果中提取可复用的洞见，供 OuroborosLoop 更新吸引子。
 */

export interface DreamFact {
  id: string;
  topic: string;
  fact: string;
  confidence: number;
  timestamp: number;
}

export interface DreamInsight {
  id: string;
  dreamType: 'replay' | 'shuffle' | 'ladder' | 'anomaly' | 'water';
  content: string;
  sourceFacts: string[];
  confidence: number;
  verified: boolean;
  createdAt: string;
}

export interface DreamCycleResult {
  consolidated: number;
  blended: number;
  abstracted: number;
  anomalies: number;
  waterPatterns: number;
  emergentInsights: DreamInsight[];
}

function _genId(): string {
  return `${Date.now().toString(36)}-${Math.random().toString(36).slice(2, 8)}`;
}

function _now(): string {
  return new Date().toISOString();
}

export class DreamConsolidator {
  private _facts: DreamFact[] = [];
  private _insights: DreamInsight[] = [];
  private _dreamCount = 0;
  private _lastDream = _now();
  private _cycles: DreamCycleResult[] = [];

  get facts(): DreamFact[] {
    return [...this._facts];
  }

  get insights(): DreamInsight[] {
    return [...this._insights];
  }

  get dreamCount(): number {
    return this._dreamCount;
  }

  ingestFact(fact: Omit<DreamFact, 'id' | 'timestamp'>): DreamFact {
    const full: DreamFact = {
      ...fact,
      id: `fact-${_genId()}`,
      timestamp: Date.now(),
    };
    this._facts.push(full);
    if (this._facts.length > 200) this._facts = this._facts.slice(-100);
    return full;
  }

  ingestAssessment(
    iteration: number,
    quality: number,
    waterPassed: boolean,
    violations: string[],
  ): DreamCycleResult {
    this.ingestFact({
      topic: 'assessment',
      fact: `iteration ${iteration}: quality=${quality.toFixed(3)}, water=${waterPassed}`,
      confidence: quality,
    });
    if (!waterPassed && violations.length) {
      for (const v of violations.slice(0, 3)) {
        this.ingestFact({
          topic: 'water-logic',
          fact: `water violation: ${v}`,
          confidence: 0.5,
        });
      }
    }
    return this.dreamCycle();
  }

  dreamCycle(): DreamCycleResult {
    this._dreamCount++;
    const result: DreamCycleResult = {
      consolidated: 0,
      blended: 0,
      abstracted: 0,
      anomalies: 0,
      waterPatterns: 0,
      emergentInsights: [],
    };

    result.consolidated = this._consolidateFacts();
    result.blended = this._shuffleBlend();
    result.abstracted = this._ladderAbstraction();
    result.anomalies = this._detectAnomalies();
    result.waterPatterns = this._extractWaterPatterns();

    result.emergentInsights = this._insights
      .filter(i => !i.verified)
      .sort((a, b) => b.confidence - a.confidence)
      .slice(0, 5);

    this._cycles.push(result);
    if (this._cycles.length > 50) this._cycles = this._cycles.slice(-30);
    this._lastDream = _now();
    return result;
  }

  verifyInsight(insightId: string, verified = true): void {
    const insight = this._insights.find(i => i.id === insightId);
    if (!insight) return;
    insight.verified = verified;
    insight.confidence = verified
      ? Math.min(1.0, insight.confidence + 0.3)
      : Math.max(0.2, insight.confidence - 0.2);
  }

  getPendingInsights(topK = 3): DreamInsight[] {
    return this._insights
      .filter(i => !i.verified)
      .sort((a, b) => b.confidence - a.confidence)
      .slice(0, topK);
  }

  stats(): Record<string, unknown> {
    return {
      dream_count: this._dreamCount,
      insights_total: this._insights.length,
      insights_unverified: this._insights.filter(i => !i.verified).length,
      last_dream: this._lastDream,
      recent_cycles: this._cycles.slice(-5),
    };
  }

  private _consolidateFacts(): number {
    // 提升高频事实的置信度，衰减长期未使用的事实
    const now = Date.now();
    for (const fact of this._facts) {
      const age = (now - fact.timestamp) / 1000;
      if (age < 60) {
        fact.confidence = Math.min(1.0, fact.confidence + 0.02);
      } else if (age > 300) {
        fact.confidence = Math.max(0.1, fact.confidence - 0.01);
      }
    }
    return Math.min(this._facts.length, 5);
  }

  private _shuffleBlend(): number {
    if (this._facts.length < 4) return 0;
    const byTopic = new Map<string, DreamFact[]>();
    for (const f of this._facts) {
      const list = byTopic.get(f.topic) ?? [];
      list.push(f);
      byTopic.set(f.topic, list);
    }
    const topics = Array.from(byTopic.keys());
    if (topics.length < 2) return 0;

    let blended = 0;
    for (let i = 0; i < Math.min(3, topics.length - 1); i++) {
      const [t1, t2] = this._sampleTopics(topics);
      const f1 = this._sample(byTopic.get(t1) ?? []);
      const f2 = this._sample(byTopic.get(t2) ?? []);
      if (!f1 || !f2 || f1.id === f2.id) continue;
      const score = (f1.confidence + f2.confidence) / 2;
      if (score > 0.55) {
        const insight: DreamInsight = {
          id: `dream-${this._dreamCount}-${blended + 1}`,
          dreamType: 'shuffle',
          content: `跨领域融合：${f1.topic}中的「${this._extractCore(f1.fact)}」与${f2.topic}中的「${this._extractCore(f2.fact)}」可能共享底层模式。`,
          sourceFacts: [f1.fact, f2.fact],
          confidence: score,
          verified: false,
          createdAt: _now(),
        };
        this._insights.push(insight);
        blended++;
      }
    }
    this._trimInsights();
    return blended;
  }

  private _ladderAbstraction(): number {
    const byTopic = new Map<string, DreamFact[]>();
    for (const f of this._facts) {
      const list = byTopic.get(f.topic) ?? [];
      list.push(f);
      byTopic.set(f.topic, list);
    }

    let abstracted = 0;
    for (const [topic, list] of byTopic) {
      if (list.length < 3) continue;
      const text = list.slice(-6).map(f => f.fact).join(' ');
      const keywords = this._extractKeywords(text);
      if (keywords.length >= 2) {
        const insight: DreamInsight = {
          id: `ladder-${this._dreamCount}-${abstracted + 1}`,
          dreamType: 'ladder',
          content: `在${topic}领域，${keywords[0]}与${keywords[1]}构成核心机制，共有的底层模式是${this._inferModality(topic, keywords)}。`,
          sourceFacts: list.slice(0, 4).map(f => f.fact),
          confidence: 0.55,
          verified: false,
          createdAt: _now(),
        };
        this._insights.push(insight);
        abstracted++;
      }
    }
    this._trimInsights();
    return abstracted;
  }

  private _detectAnomalies(): number {
    let anomalies = 0;
    for (let i = 0; i < this._facts.length && anomalies < 5; i++) {
      for (let j = i + 1; j < this._facts.length && anomalies < 5; j++) {
        const f1 = this._facts[i];
        const f2 = this._facts[j];
        if (f1.topic === f2.topic && this._areContradictory(f1.fact, f2.fact)) {
          const insight: DreamInsight = {
            id: `anomaly-${this._dreamCount}-${anomalies + 1}`,
            dreamType: 'anomaly',
            content: `潜在矛盾：[${f1.fact}] vs [${f2.fact}] — 需确认`,
            sourceFacts: [f1.fact, f2.fact],
            confidence: 0.35,
            verified: false,
            createdAt: _now(),
          };
          this._insights.push(insight);
          anomalies++;
        }
      }
    }
    this._trimInsights();
    return anomalies;
  }

  private _extractWaterPatterns(): number {
    const waterFacts = this._facts.filter(f => f.topic === 'water-logic');
    if (waterFacts.length < 2) return 0;
    const keywords = this._extractKeywords(waterFacts.map(f => f.fact).join(' '));
    const top = keywords.slice(0, 2).join('、') || '动量/能量异常';
    const insight: DreamInsight = {
      id: `water-${this._dreamCount}-1`,
      dreamType: 'water',
      content: `水逻辑反复违规模式：${top}。建议下一轮增强物理约束或在生成阶段降低相关自由度。`,
      sourceFacts: waterFacts.slice(-4).map(f => f.fact),
      confidence: 0.6,
      verified: false,
      createdAt: _now(),
    };
    this._insights.push(insight);
    this._trimInsights();
    return 1;
  }

  private _sample<T>(arr: T[]): T | undefined {
    if (!arr.length) return undefined;
    return arr[Math.floor(Math.random() * arr.length)];
  }

  private _sampleTopics(topics: string[]): [string, string] {
    let a = this._sample(topics)!;
    let b = this._sample(topics)!;
    while (topics.length > 1 && a === b) {
      b = this._sample(topics)!;
    }
    return [a, b];
  }

  private _extractCore(fact: string): string {
    const match = fact.match(/[\u4e00-\u9fa5a-zA-Z]{2,16}/);
    return match ? match[0].trim() : '概念';
  }

  private _extractKeywords(text: string): string[] {
    const words = (text.match(/[\u4e00-\u9fa5a-zA-Z]{2,12}/g) ?? []);
    const freq = new Map<string, number>();
    for (const w of words) {
      freq.set(w, (freq.get(w) ?? 0) + 1);
    }
    return Array.from(freq.entries())
      .filter(([, c]) => c >= 2)
      .sort((a, b) => b[1] - a[1])
      .slice(0, 5)
      .map(([w]) => w);
  }

  private _inferModality(_topic: string, keywords: string[]): string {
    const patterns: Record<string, string> = {
      '优化': '基于目标函数的最小化原则',
      '学习': '从经验中归纳出可泛化的规则',
      '结构': '层级组织的模块化设计',
      '演化': '变异+选择的迭代过程',
      '计算': '符号转换与状态迁移',
      '信息': '不确定性的消除与传递',
      '网络': '节点间的连接与信号传播',
    };
    for (const kw of keywords) {
      for (const [key, value] of Object.entries(patterns)) {
        if (kw.includes(key)) return value;
      }
    }
    return '自组织涌现';
  }

  private _areContradictory(f1: string, f2: string): boolean {
    const pairs = [
      ['是', '不是'],
      ['可以', '不可以'],
      ['能', '不能'],
      ['属于', '不属于'],
      ['支持', '不支持'],
    ];
    for (const [pos, neg] of pairs) {
      if ((f1.includes(pos) && f2.includes(neg)) || (f1.includes(neg) && f2.includes(pos))) {
        return true;
      }
    }
    return false;
  }

  private _trimInsights(): void {
    if (this._insights.length > 200) {
      this._insights = this._insights.slice(-100);
    }
  }
}
